### PR TITLE
[IA-3338] Fix existing pd + new runtime attaching issue

### PR DESCRIPTION
**Pull request by jdcanas**
The core issue is that the UI was deciding to include just a `diskSize` when a persistent disk existed and a `name` within a `persistentDisk` payload was necessary. The fix was to separate the `runtimeConfig` objects disk consideration into the existing if statements for update/create.

We re-created this issue by performing the following on dev before the code changes
1. making a runtime with a PD in a new workspace
2. deleting the runtime but not the pd,
3. creating a new runtime, and verifying the createRuntime payload had a diskSize, and no object of the form `{ persistentDisk: { name: ... } }`. In other words, the UI did not recognize the existing disk and created a new one

To verify the fix
1. Perform 1-2 from above, and take note of the pd name in the `runtimeConfig` in the network tab
2. Create a new runtime, and verify the payload has a `name` field matches the existing pd
image

We also verified various things such as
- modifying the image and ensuring the pd is passed along to the new runtime
- modifying pd size and ensuring update behaves as expected
- changing from a gce to dataproc runtime (this actually uncovered an existing bug, ticket created here https://broadworkbench.atlassian.net/browse/IA-3338)
- verifying users are prompted to upgrade to pd when they have a pd (a banner at the bottom of the compute modal)

### CODE REVIEW

**Review by Qi77Qi — COMMENTED** [review]
is this sth we can unit test?

File: src/components/ComputeModal.js [hunk 1]
```diff
@@ -309,24 +277,59 @@ export const ComputeModalBase = ({
...
+          ...(desiredRuntime.numberOfWorkers && {
+            numberOfPreemptibleWorkers: desiredRuntime.numberOfPreemptibleWorkers,
+            workerMachineType: desiredRuntime.workerMachineType,
+            workerDiskSize: desiredRuntime.workerDiskSize
+          })
+        })
+      }
+
+      if (shouldUpdateRuntime) {
+        const updateRuntimeConfig = _.merge(shouldUpdatePersistentDisk ? { diskSize: desiredPersistentDisk.size } : {}, runtimeConfig)
```
Comment (kyuksel) [inline src/components/ComputeModal.js:303]: Does this need to be a deep merge or would shallow merge suffice (e.g. using the spread operator)?

**Review by kyuksel — APPROVED** [review]
Code LGTM with one consideration inline.
Didn't test.

File: src/components/ComputeModal.js (hunk 1 above)
Comment (jdcanas) [reply src/components/ComputeModal.js:303]: The spread operator would work, but it looks a bit cleaner this way imo. I'm not tied to it though

**Comment by jdcanas** [comment]
> is this sth we can unit test?

Not very easily... it's dependent on a lot of state and other functions. It might be worth having an integration test though, to ensure if I delete my runtime and create a new one my disk is attached. Should be verifiable via leo APIs.

**Review by gpcarr — APPROVED** [review]
lgtm

File: src/components/ComputeModal.js (hunk 1 above)
Comment (kyuksel) [reply src/components/ComputeModal.js:303]: I like the way it looks too; just thought we could eliminate unnecessary operations and adhere to the rule of least power but I don't feel strongly.

**Comment by gpcarr** [comment]
tested by doing the following:
```
// original create runtime payload on branch dev (we'll call this runtime1):
{
    "runtimeConfig": {
        "cloudService": "GCE",
        "zone": "us-central1-a",
        "machineType": "n1-standard-1",
        "persistentDisk": {
            "name": "saturn-pd-9735c6a9-007e-474c-becd-1f11520d9cb5",
            "size": 50,
            "diskType": "pd-standard",
            "labels": {
                "saturnWorkspaceNamespace": "general-dev-billing-account",
                "saturnWorkspaceName": "final-test"
            }
        }
    },
    "autopauseThreshold": 30,
    "toolDockerImage": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:2.0.8",
    "labels": {
        "saturnWorkspaceNamespace": "general-dev-billing-account",
        "saturnWorkspaceName": "final-test",
        "saturnAutoCreated": "true",
        "saturnVersion": "6"
    },
    "customEnvironmentVariables": {
        "WORKSPACE_NAME": "final-test",
        "WORKSPACE_NAMESPACE": "general-dev-billing-account",
        "WORKSPACE_BUCKET": "gs://fc-3e9b036f-5be0-40fc-8ed4-94de679961b5",
        "GOOGLE_PROJECT": "terra-dev-65774f60"
    },
    "defaultClientId": "500025638838-s2v23ar3spugtd5t2v1vgfa2sp7ppg0d.apps.googleusercontent.com",
    "userJupyterExtensionConfig": {
        "nbExtensions": {
            "saturn-iframe-extension": "https://bvdp-saturn-dev.appspot.com/jupyter-iframe-extension.js"
        },
        "labExtensions": {},
        "serverExtensions": {},
        "combinedExtensions": {}
    },
    "scopes": [
        "https://www.googleapis.com/auth/cloud-platform",
        "https://www.googleapis.com/auth/userinfo.email",
        "https://www.googleapis.com/auth/userinfo.profile"
    ],
    "enableWelder": true
}

// do a delete runtime but keep PD
// recreate a new jupyter runtime (we'll call this runtime2), i expect the disk to be saturn-pd-9735c6a9-007e-474c-becd-1f11520d9cb5
// second create payload (still on branch dev):
{
    "runtimeConfig": {
        "cloudService": "GCE",
        "zone": "us-central1-a",
        "machineType": "n1-standard-1",
        "diskSize": 50
    },
    "autopauseThreshold": 30,
    "toolDockerImage": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:2.0.8",
    "labels": {
        "saturnWorkspaceNamespace": "general-dev-billing-account",
        "saturnWorkspaceName": "final-test",
        "saturnAutoCreated": "true",
        "saturnVersion": "6"
    },
    "customEnvironmentVariables": {
        "WORKSPACE_NAME": "final-test",
        "WORKSPACE_NAMESPACE": "general-dev-billing-account",
        "WORKSPACE_BUCKET": "gs://fc-3e9b036f-5be0-40fc-8ed4-94de679961b5",
        "GOOGLE_PROJECT": "terra-dev-65774f60"
    },
    "defaultClientId": "500025638838-s2v23ar3spugtd5t2v1vgfa2sp7ppg0d.apps.googleusercontent.com",
    "userJupyterExtensionConfig": {
        "nbExtensions": {
            "saturn-iframe-extension": "https://bvdp-saturn-dev.appspot.com/jupyter-iframe-extension.js"
        },
        "labExtensions": {},
        "serverExtensions": {},
        "combinedExtensions": {}
    },
    "scopes": [
        "https://www.googleapis.com/auth/cloud-platform",
        "https://www.googleapis.com/auth/userinfo.email",
        "https://www.googleapis.com/auth/userinfo.profile"
    ],
    "enableWelder": true
}
// runtimeConfig for runtime2 does not pass a disk name
// runtime2 GET:
{
    "id": 67400,
    "runtimeName": "saturn-d3d3b55c-f591-42e6-a35d-e135be19cebd",
    "googleProject": "terra-dev-65774f60",
    "cloudContext": {
        "cloudProvider": "GCP",
        "cloudResource": "terra-dev-65774f60"
    },
    "serviceAccount": "pet-100271339377034276912@terra-dev-65774f60.iam.gserviceaccount.com",
    "asyncRuntimeFields": {
        "googleId": "b2975454f4836bcc",
        "operationName": "operation-1650645056890-5dd40c0442666-984e1b3d-9dd43742:terra-dev-65774f60:us-central1-a",
        "stagingBucket": "leostaging-saturn-d3d3b55c-d3738bbc-01f0-44de-ad2f-f29485e29ca4",
        "hostIp": "34.135.21.41"
    },
    "auditInfo": {
        "creator": "b.adm.firec@gmail.com",
        "createdDate": "2022-04-22T16:30:49.808785Z",
        "destroyedDate": null,
        "dateAccessed": "2022-04-22T16:31:32.703Z"
    },
    "runtimeConfig": {
        "machineType": "n1-standard-1",
        "diskSize": 50,
        "cloudService": "GCE",
        "bootDiskSize": 100,
        "zone": "us-central1-a",
        "gpuConfig": null
    },
    "proxyUrl": "https://leonardo.dsde-dev.broadinstitute.org/proxy/terra-dev-65774f60/saturn-d3d3b55c-f591-42e6-a35d-e135be19cebd/jupyter",
    "status": "Creating",
    "labels": {
        "saturnWorkspaceNamespace": "general-dev-billing-account",
        "saturn-iframe-extension": "https://bvdp-saturn-dev.appspot.com/jupyter-iframe-extension.js",
        "creator": "b.adm.firec@gmail.com",
        "clusterServiceAccount": "pet-100271339377034276912@terra-dev-65774f60.iam.gserviceaccount.com",
        "saturnAutoCreated": "true",
        "clusterName": "saturn-d3d3b55c-f591-42e6-a35d-e135be19cebd",
        "saturnWorkspaceName": "final-test",
        "saturnVersion": "6",
        "tool": "Jupyter",
        "runtimeName": "saturn-d3d3b55c-f591-42e6-a35d-e135be19cebd",
        "cloudContext": "Gcp/terra-dev-65774f60",
        "googleProject": "terra-dev-65774f60"
    },
    "userScriptUri": null,
    "startUserScriptUri": null,
    "jupyterUserScriptUri": null,
    "jupyterStartUserScriptUri": null,
    "errors": [],
    "userJupyterExtensionConfig": {
        "nbExtensions": {
            "saturn-iframe-extension": "https://bvdp-saturn-dev.appspot.com/jupyter-iframe-extension.js"
        },
        "serverExtensions": {},
        "combinedExtensions": {},
        "labExtensions": {}
    },
    "autopauseThreshold": 30,
    "defaultClientId": "500025638838-s2v23ar3spugtd5t2v1vgfa2sp7ppg0d.apps.googleusercontent.com",
    "runtimeImages": [
        {
            "imageType": "Welder",
            "imageUrl": "us.gcr.io/broad-dsp-gcr-public/welder-server:72e1f03",
            "homeDirectory": null,
            "timestamp": "2022-04-22T16:30:49.808785Z"
        },
        {
            "imageType": "BootSource",
            "imageUrl": "projects/broad-dsp-gcr-public/global/images/custom-leo-gce-cos-image-aa373a8",
            "homeDirectory": null,
            "timestamp": "2022-04-22T16:30:52.070Z"
        },
        {
            "imageType": "Proxy",
            "imageUrl": "broadinstitute/openidc-proxy:2.3.1_2",
            "homeDirectory": null,
            "timestamp": "2022-04-22T16:30:49.808785Z"
        },
        {
            "imageType": "CryptoDetector",
            "imageUrl": "us.gcr.io/broad-dsp-gcr-public/cryptomining-detector:0.0.2",
            "homeDirectory": null,
            "timestamp": "2022-04-22T16:30:49.808785Z"
        },
        {
            "imageType": "Jupyter",
            "imageUrl": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:2.0.8",
            "homeDirectory": "/home/jupyter",
            "timestamp": "2022-04-22T16:30:49.808785Z"
        }
    ],
    "scopes": [
        "https://www.googleapis.com/auth/cloud-platform",
        "https://www.googleapis.com/auth/userinfo.email",
        "https://www.googleapis.com/auth/userinfo.profile"
    ],
    "customEnvironmentVariables": {
        "WORKSPACE_NAME": "final-test",
        "WORKSPACE_NAMESPACE": "general-dev-billing-account",
        "WORKSPACE_BUCKET": "gs://fc-3e9b036f-5be0-40fc-8ed4-94de679961b5",
        "GOOGLE_PROJECT": "terra-dev-65774f60"
    },
    "diskConfig": null,
    "patchInProgress": false
}
// check google console to see what PD is mounted to runtime2
// PD in console that is attached = saturn-d3d3b55c-f591-42e6-a35d-e135be19cebd-1 which is NOT the user's original PD
// check in cloud environments page to see the original PD: saturn-pd-9735c6a9-007e-474c-becd-1f11520d9cb5
// switch to justin's branch jc-fix-size-error-on-delete NOTE: i had to restart my local terra-ui server for the change to work
// delete runtime2 that was created as non PD (the runtime that did not get the right PD attached to it")
// choose the first option for disks when deleting ("Delete application configuration and cloud compute profile)
// see that runtime2 and the disk that got created with it get deleted in google (ie not the original PD but rather saturn-d3d3b55c-f591-42e6-a35d-e135be19cebd-1)
// confirm in cloud environments page that original PD (saturn-pd-9735c6a9-007e-474c-becd-1f11520d9cb5) is still there
// create a new runtime (we'll call this runtime3)
// POST for create runtime3
{
    "runtimeConfig": {
        "cloudService": "GCE",
        "zone": "us-central1-a",
        "machineType": "n1-standard-1",
        "persistentDisk": {
            "name": "saturn-pd-9735c6a9-007e-474c-becd-1f11520d9cb5"
        }
    },
    "autopauseThreshold": 30,
    "toolDockerImage": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:2.0.8",
    "labels": {
        "saturnWorkspaceNamespace": "general-dev-billing-account",
        "saturnWorkspaceName": "final-test",
        "saturnAutoCreated": "true",
        "saturnVersion": "6"
    },
    "customEnvironmentVariables": {
        "WORKSPACE_NAME": "final-test",
        "WORKSPACE_NAMESPACE": "general-dev-billing-account",
        "WORKSPACE_BUCKET": "gs://fc-3e9b036f-5be0-40fc-8ed4-94de679961b5",
        "GOOGLE_PROJECT": "terra-dev-65774f60"
    },
    "defaultClientId": "500025638838-s2v23ar3spugtd5t2v1vgfa2sp7ppg0d.apps.googleusercontent.com",
    "userJupyterExtensionConfig": {
        "nbExtensions": {
            "saturn-iframe-extension": "https://bvdp-saturn-dev.appspot.com/jupyter-iframe-extension.js"
        },
        "labExtensions": {},
        "serverExtensions": {},
        "combinedExtensions": {}
    },
    "scopes": [
        "https://www.googleapis.com/auth/cloud-platform",
        "https://www.googleapis.com/auth/userinfo.email",
        "https://www.googleapis.com/auth/userinfo.profile"
    ],
    "enableWelder": true
}
// with the fix we see that persistentDisk.name gets correctly passed which is the origial older disk saturn-pd-9735c6a9-007e-474c-becd-1f11520d9cb5
// we also see the message:
// "Time to upgrade your cloud environment. Terra’s new persistent disk feature will safeguard your work and data. Learn more about Persistent disks and where your disk is mounted"
// go away since the UI is now correctly attaching the original PD
// double check in google console that runtime3 has the original PD (saturn-pd-9735c6a9-007e-474c-becd-1f11520d9cb5) attached to it 
```

**Pull request merged**